### PR TITLE
New tab button

### PR DESCRIPTION
**Pull request by wldcordeiro**
Associated Issue: #951 

### Summary of Changes

* Add a new tab button that opens the file search
* move file search toggle function up into App.js so that it could be shared between `SourceTabs` and `SourceSearch`

### Testing

* [x] passes `npm test`
* [x] passes `npm run lint`

### Screenshots/Videos (OPTIONAL)

screenshot_20161025_213239
screenshot_20161025_213248
screenshot_20161025_213254
screenshot_20161025_213720

### CODE REVIEW

File: public/js/components/SourceSearch.js [hunk 1]
```diff
@@ -61,27 +57,27 @@ const Search = React.createClass({
...
 
   render() {
-    return this.state.searchOn ?
+    return this.props.searchOn ?
       dom.div({ className: "search-container" },
       Autocomplete({
         selectItem: result => {
           this.props.selectSource(result.id);
-          this.setState({ searchOn: false });
+          this.props.toggleSearch(false);
```
Comment (jasonLaster) [inline public/js/components/SourceSearch.js]: Hmm, I'm not thrilled about refactoring SourcesState so that the searchOn state is kept in the app. I can see the argument for it being app state, but at this point we have a nice convention for state being local to the component it is being used.

I think it would be nice to use DOM events to open the panel. This way, the panel API is largely DOM and doesn't rely on React.

Here is a function for opening the panel that could be use in the tab onClick.
```js
function openSearchPanel() {
  window.dispatchEvent(Object.assign(new Event("keydown"), {
    key: "p", metaKey: true, ctrlKey: false, altKey: false, shiftKey: false
  }))
}
```

File: public/js/components/SourceTabs.js [hunk 2]
```diff
@@ -139,6 +140,13 @@ const SourceTabs = React.createClass({
       CloseButton({ handleClick: onClickClose }));
   },
 
+  renderNewButton() {
+    return dom.div({
+      className: "new-tab-btn",
+      onClick: () => this.props.openFileSearch(true)
```
Comment (jasonLaster) [inline public/js/components/SourceTabs.js]: so this wasn't clear because of the dark theme tabs, but is it possible to have multiple blank tabs?

It would be nice if the tabs button were hidden during file search.

Comment (jasonLaster) [inline public/js/components/SourceTabs.js]: ahh - good point. I guess the next best thing to do would be to make the button a no-op when file search is open.

File: public/js/components/SourceSearch.js (hunk 1 above)
Comment (wldcordeiro) [inline public/js/components/SourceSearch.js]: I'm not really a fan of having a function that will be bound to an `onClick` handler that then fires a `keydown` event. Maybe this is something that should live in a Redux reducer.

Comment (jasonLaster) [inline public/js/components/SourceSearch.js]: Hmm, lets see what the reducer option looks like. We do want a UI state reducer at some point. 

I'm on the fence about the DOM event. I think i'd also feel differently if instead of it being a keydown it were clicking an element :)

**Comment by jasonLaster** [comment]
@jlongster mind giving this a look as it introduces a new UI reducer.

File: public/js/actions/types.js [hunk 3]
```diff
@@ -88,7 +87,7 @@ type SourceAction =
                sourceText: SourceText }}
   | { type: "CLOSE_TAB", id: string };
 
-type UIAction = { type: constants.TOGGLE_FILE_SEARCH, searchOn: boolean };
+type UIAction = { type: "TOGGLE_FILE_SEARCH", searchOn: boolean };
```
Comment (wldcordeiro) [inline public/js/actions/types.js]: Yeah, good idea. This file could likely be updated to use the constants in its own PR.

File: public/js/actions/ui.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// @flow
+const constants = require("../constants");
+import type { ThunkArgs } from "./types";
+
+function toggleFileSearch(searchOn: boolean) {
+  return ({ dispatch }: ThunkArgs) => {
+    dispatch({
+      type: constants.TOGGLE_FILE_SEARCH,
+      searchOn
+    });
```
Comment (jlongster) [inline public/js/actions/ui.js]: You don't need to use the thunk middleware here, just return the action:

```
function toggleFileSearch(searchOn: boolean) {
  return {
      type: constants.TOGGLE_FILE_SEARCH,
      searchOn
  };
}

File: public/js/actions/types.js (hunk 3 above)
Comment (jlongster) [inline public/js/actions/types.js]: It doesn't need to use the constants. Once everything is flow typed the constants file can go away. Flow will error if you pass in a string that doesn't match any of these types.

**Comment by jlongster** [comment]
Code looks great! I haven't tested this locally, so as long as it works, the code is fine. What happens when there are too many tabs and we start hiding them? etc Test all the edge cases manually and make sure nothing is broken.

**Comment by jasonLaster** [comment]
@jlongster you're right that the small width UI challenges are the difficult edge case.

Here's a gif of it working

**Comment by jasonLaster** [comment]
Thanks @wldcordeiro for this great work. I really appreciate how you jump in on the UI polish tickets.

Hope my small tweaks are alright.

**Comment by wldcordeiro** [comment]
@jasonLaster looks like we may want to set a minimum width for the center pane? The search results break out of their container in an ugly way too.

**Comment by jasonLaster** [comment]
@wldcordeiro that's a good idea. I'll file an issue for that.

**Comment by clarkbw** [comment]
> I'll file an issue for that.

#1029 is handling the minimum width
